File: functions/ruby/functions-Jeremy.rb
#!/usr/bin/env ruby
#
# Introduction to Functions in ruby

# Python and ruby have very simular features, first being
# default arguments
def addFive(num=5)
  return 5 + num
end

puts "First no arguments #{addFive()}"
puts "Now with an argument #{addFive(9)}"

# Lambdas are also in ruby
def subtractClosure(num)
  return lambda {|x| num - x}
end

clo = subtractClosure(10)

puts "First call: #{clo[5]}"
puts "Second call: #{clo[2]}"

# Reset the closure
clo = subtractClosure(5)
puts "(new)First call: #{clo[5]}"
puts "(new)Second call: #{clo[2]}"

File: functions/python/functions-Jeremy.py
#!/usr/bin/env python
#
# Introduction to Functions in python

# Python can have both parameters and default arguments
# inside the parentheses
def printNum(num=64):
  print 'Your number is,', num

# No arguments will always be 64
printNum()

# With an argument the default changes
printNum(44)

# Python have anonymous functions or lambdas so you can
# make closures just like in javascript
def addClosure(num):
  return lambda x: num + x

# The function is put into this variable
clo = addClosure(5)

# the number is now forever captured and can be used
# with the new function
print 'First call:', clo(2)
print 'Second call:', clo(8)

# Change the closure variable
clo = addClosure(10)

print '(new)First call:', clo(2)
print '(new)Second call:', clo(8)